pkg/scuzz/ata.go

```diff
@@ -0,0 +1,281 @@
+	readVerify            Cmd = 0x40
+	readVerifyOnce        Cmd = 0x41
+	readVerifyExt         Cmd = 0x42
+	writeUNCExt           Cmd = 0x45 // lba48 no data uses feat reg
+	formatTrack           Cmd = 0x50
+	downloadMicrocode     Cmd = 0x92
+	standbynow2           Cmd = 0x94
+	checkpowermode2       Cmd = 0x98
+	sleepnow2             Cmd = 0x99
+	pidentify             Cmd = 0xa1
+	readNativeMax         Cmd = 0xf8
+	readNativeMaxExt      Cmd = 0x27
+	smart                 Cmd = 0xb0
+	dco                   Cmd = 0xb1
+	eraseSectors          Cmd = 0xc0
+	readDMA               Cmd = 0xc8
+	writeDMA              Cmd = 0xca
+	doorlock              Cmd = 0xde
+	doorunlock            Cmd = 0xdf
+	standbynow1           Cmd = 0xe0
+	idleimmediate         Cmd = 0xe1
+	setidle               Cmd = 0xe3
+	setMax                Cmd = 0xf9
+	setMaxExt             Cmd = 0x37
+	setMultiple           Cmd = 0xc6
+	checkpowermode1       Cmd = 0xe5
+	sleepnow1             Cmd = 0xe6
+	flushcache            Cmd = 0xe7
+	flushcacheExt         Cmd = 0xea
+	identify              Cmd = 0xec
+	setfeatures           Cmd = 0xef
+	securitySetPass       Cmd = 0xf1
+	// SecurityUnlock unlocks the drive with a given 32-byte password
+	SecurityUnlock       Cmd = 0xf2
+	securityErasePrepare Cmd = 0xf3
+	securityEraseUnit    Cmd = 0xf4
+	securityFreezeLock   Cmd = 0xf5
+	securityDisable      Cmd = 0xf6
+	vendorSpecific0x80   Cmd = 0x80
+)
+
+// This is a map of commands to common command properties.
+// For most of these properties we don't know yet how to set them
+// so this will have to grow over time.
+var properties = map[Cmd]struct {
+	cdb cdb
+}{
+	dsm:                  {},
+	readPIO:              {},
+	readPIOOnce:          {},
+	readLong:             {},
+	readLongOnce:         {},
+	readPIOExt:           {},
+	readDMAExt:           {},
+	readFPDMA:            {},
+	writePIO:             {},
+	writeLong:            {},
+	writeLongOnce:        {},
+	writePIOExt:          {},
+	writeDMAExt:          {},
+	writeFPDMA:           {},
+	readVerify:           {},
+	readVerifyOnce:       {},
+	readVerifyExt:        {},
+	writeUNCExt:          {},
+	formatTrack:          {},
+	downloadMicrocode:    {},
+	standbynow2:          {},
+	checkpowermode2:      {},
+	sleepnow2:            {},
+	pidentify:            {},
+	readNativeMax:        {},
+	readNativeMaxExt:     {},
+	smart:                {},
+	dco:                  {},
+	eraseSectors:         {},
+	readDMA:              {},
+	writeDMA:             {},
+	doorlock:             {},
+	doorunlock:           {},
+	standbynow1:          {},
+	idleimmediate:        {},
+	setidle:              {},
+	setMax:               {},
+	setMaxExt:            {},
+	setMultiple:          {},
+	checkpowermode1:      {},
+	sleepnow1:            {},
+	flushcache:           {},
+	flushcacheExt:        {},
+	identify:             {},
+	setfeatures:          {},
+	securitySetPass:      {},
+	SecurityUnlock:       {cdb: cdb{ata16, lba48 | pioOut}},
```

Comment:
this should use commandDataBlock and whatever uses this array should use .marshal()?

Reply:
I am pretty sure not. I am hoping these can be 5 byte slices that form the first part of the marshalled command block, then the rest of it gets filled out from the struct. I'm not going to know until I get more commands implemented however. So a marshal takes the opcode, grabs the first five bytes here, then fills out the rest based on the block number and other bits.